vm/src/warn.rs

```diff
@@ -55,8 +197,35 @@ fn warn_explicit(
     } else {
         vm.ctx.exceptions.user_warning.to_owned()
     };
+
+    // Create key.
+    let key = PyTuple::new_ref(
+        vec![
+            vm.ctx.new_int(3).into(),
+            vm.ctx.new_str(text).into(),
+            category.as_object().to_owned(),
+            vm.ctx.new_int(lineno).into(),
+        ],
+        &vm.ctx,
+    );
+
+    if already_warned(registry, key.as_object().to_owned(), 0, vm) {
+        return Ok(());
+    }
```

Comment:
We only need to check already_warned when registry is not `None`.
```suggestion
    if !vm.is_none(registry.as_object()) && already_warned(registry, key.as_object().to_owned(), 0, vm) {
        return Ok(());
    }
```